src/predictor/gpu_predictor.cu

```diff
@@ -270,26 +246,37 @@ class GPUPredictor : public xgboost::Predictor {
     InitModel(model, h_tree_segments, h_nodes, tree_begin, tree_end);
   }
 
-  void DevicePredictInternal(DMatrix* dmat,
-                             HostDeviceVector<bst_float>* out_preds,
+  void DevicePredictInternal(DMatrix* dmat, HostDeviceVector<float>* out_preds,
                              const gbm::GBTreeModel& model, size_t tree_begin,
-                             size_t tree_end) {
+                             size_t tree_end, bool precise) {
+    dh::safe_cuda(cudaSetDevice(generic_param_->gpu_id));
     if (tree_end - tree_begin == 0) {
       return;
     }
     monitor_.StartCuda("DevicePredictInternal");
-
     InitModel(model, tree_begin, tree_end);
+    out_preds->SetDevice(generic_param_->gpu_id);
 
-    size_t batch_offset = 0;
-    for (auto &batch : dmat->GetBatches<SparsePage>()) {
-      batch.offset.SetDevice(generic_param_->gpu_id);
-      batch.data.SetDevice(generic_param_->gpu_id);
-      PredictInternal(batch, model.learner_model_param_->num_feature,
-                      out_preds, batch_offset);
-      batch_offset += batch.Size() * model.learner_model_param_->num_output_group;
-    }
+    auto do_predict_sparse_page = [&]() {
+      size_t batch_offset = 0;
+      for (auto &batch : dmat->GetBatches<SparsePage>()) {
+        this->PredictInternal(batch, model.learner_model_param_->num_feature,
+                              out_preds, batch_offset);
+        batch_offset += batch.Size() * model.learner_model_param_->num_output_group;
+      }
+    };
 
+    if (precise && dmat->PageExists<SparsePage>()) {
+      do_predict_sparse_page();
+    } else if (!dmat->PageExists<EllpackPage>()) {
```

Comment:
We need to make sure this EllpackPage exists at the time of the first prediction call, otherwise the SparsePage will be created on device in the first iteration.

Reply:
That's the part where I need to know the plan for DMatrix initialization.